### PR TITLE
Upgrade the Rust & Clippy versions

### PR DESCRIPTION
**Pull request by Hoverbear**
This PR updates the Rust version to `nightly-2018-04-06` and the `clippy` and `clippy_lint` versions to `0.0.192` which match with the `nightly-2018-04-04` version. This also bumps `rustfmt`.

The massive changeset is primarily due to fixing clippy lints and running `rustfmt`.

Should be merged along with https://github.com/pingcap/SRE/pull/67.

### CODE REVIEW

**Comment by ngaut** [comment]
Good job.

**Comment by siddontang** [comment]
You need to update README.md too.

Btw, do we need to upgrade rusftfmt?

**Comment by Hoverbear** [comment]
I think updating Rustfmt is a great idea!

**Comment by Hoverbear** [comment]
I did an `#![allow(nonminimal_bool)]` since it seems to have some false positive issues.

File: src/coprocessor/checksum.rs [hunk 1]
```diff
@@ -46,8 +46,8 @@ impl ChecksumContext {
             ctx.fill_cache,
         );
         ChecksumContext {
-            req: req,
-            store: store,
+            req,
```
Comment (siddontang) [inline src/coprocessor/checksum.rs]: strange that here two fields don't use `filed: value` but next two use.

File: src/coprocessor/codec/mysql/decimal.rs [hunk 2]
```diff
@@ -868,11 +870,11 @@ impl Decimal {
     /// create a new decimal for internal usage.
     fn new(int_cnt: u8, frac_cnt: u8, negative: bool) -> Decimal {
         Decimal {
-            int_cnt: int_cnt,
-            frac_cnt: frac_cnt,
+            int_cnt,
+            frac_cnt,
```
[the remark this answers is not in the document]
Comment (Hoverbear) [reply src/coprocessor/codec/mysql/decimal.rs]: I could add an `allow` flag here if you'd prefer.

File: Makefile [hunk 3]
```diff
@@ -78,7 +78,7 @@ bench:
 	RUST_BACKTRACE=1 cargo run --release --bin bench-tikv --features "${ENABLE_FEATURES}"
 
 format:
-	@cargo fmt --all -- --write-mode diff >/dev/null || cargo fmt --all
+	rustup run `tail -n 1 RUST_VERSION` cargo fmt --all -- --write-mode diff
```
[the remark this answers is not in the document]
Comment (Hoverbear) [reply Makefile]: At the moment I'm more just trying to get it to work. :(

File: README.md [hunk 4]
```diff
@@ -23,12 +23,12 @@ Thanks to the internal optimization, TiKV and TiDB can work together to be the b
 
 ### Required Rust version
 
-Rust Nightly is required. TiKV is currently tested mainly with `rust-nightly-2018-01-12`, however we would like to track `nightly`, so please report new breakage.
+Rust Nightly is required. TiKV is currently tested mainly with `rust-nightly-2018-04-06`, however we would like to track `nightly`, so please report new breakage.
```
[the remark this answers is not in the document]
Comment (Hoverbear) [reply README.md]: That's a great idea!

File: src/coprocessor/codec/datum.rs [hunk 5]
```diff
@@ -630,38 +630,38 @@ impl Datum {
...
+    pub fn checked_int_div(self, _: &mut EvalContext, datum: Datum) -> Result<Datum> {
+        match datum {
             Datum::I64(0) | Datum::U64(0) => return Ok(Datum::Null),
             _ => {}
         }
-        match (self, d) {
-            (Datum::I64(l), Datum::I64(r)) => match l.checked_div(r) {
-                None => Err(box_err!("{} intdiv {} overflow", l, r)),
+        match (self, datum) {
+            (Datum::I64(left), Datum::I64(right)) => match left.checked_div(right) {
```
[the remark this answers is not in the document]
Comment (Hoverbear) [reply src/coprocessor/codec/datum.rs]: I can add an allow flag here if you'd like, but clippy is throwing warnings for too many variables with one character names (and I agree with it).

File: docker/rust_dockerfile [hunk 6]
```diff
@@ -39,14 +39,17 @@ ENV PATH $GOPATH/bin:$GOROOT/bin:$PATH
...
 
-# cache dep package
 RUN curl -L https://github.com/pingcap/tikv/archive/master.tar.gz -o /root/master.tar.gz \
-	&& cd /root && tar xf master.tar.gz && cd tikv-master && /root/.cargo/bin/cargo fetch \
-	&& cd /root && rm -rf master.tar.gz tikv-master
+	&& cd /root && tar xf master.tar.gz \
+	&& cd /root && rm -rf master.tar.gz
+
+RUN cd /root/tikv-master \
+	&& /root/.cargo/bin/rustup override set $(tail -n 1 RUST_VERSION) \
```
[the remark this answers is not in the document]
Comment (Hoverbear) [reply docker/rust_dockerfile]: Hm, I think the behavior should be unchanged, but I'll investigate more deeply and verify this. Thanks for flagging it.

**Comment by Hoverbear** [comment]
`make` is so silly sometimes. Running `make format` works on Arch, Ubuntu, and FreeBSD, but apparently not on Circle. :-P

**Comment by breeswish** [comment]
LGTM

**Comment by breeswish** [comment]
Please resolve conflicts :P

**Comment by Hoverbear** [comment]
@breeswish Of course. :)

File: src/util/worker/future.rs [hunk 7]
```diff
@@ -172,13 +172,11 @@ impl<T: Display + Send + 'static> Worker<T> {
     pub fn stop(&mut self) -> Option<thread::JoinHandle<()>> {
         // close sender explicitly so the background thread will exit.
         info!("stoping {}", self.scheduler.name);
-        if self.handle.is_none() {
-            return None;
-        }
+        let handle = self.handle.take()?;
```
Comment (overvenus) [inline src/util/worker/future.rs]: :+1: I didn't know that!

File: docker/rust_dockerfile [hunk 8]
```diff
@@ -39,14 +39,17 @@ ENV PATH $GOPATH/bin:$GOROOT/bin:$PATH
...
 
-# cache dep package
 RUN curl -L https://github.com/pingcap/tikv/archive/master.tar.gz -o /root/master.tar.gz \
-	&& cd /root && tar xf master.tar.gz && cd tikv-master && /root/.cargo/bin/cargo fetch \
-	&& cd /root && rm -rf master.tar.gz tikv-master
+	&& cd /root && tar xf master.tar.gz \
+	&& cd /root && rm -rf master.tar.gz
+
+RUN cd /root/tikv-master \
+	&& /root/.cargo/bin/rustup default $(tail -n 1 RUST_VERSION) \
```
[the remark this answers is not in the document]
Comment (Hoverbear) [reply docker/rust_dockerfile]: Ah yes, mixed tabs and spaces, thank `vim`! :)

File: src/util/worker/future.rs (hunk 7 above)
Comment (Hoverbear) [reply src/util/worker/future.rs]: Yeah! It's a first time use for me too.

**Comment by Hoverbear** [comment]
Rebased. :)

File: docker/rust_dockerfile (hunk 6 above)
[the remark this answers is not in the document]
Comment (BusyJay) [reply docker/rust_dockerfile]: Any progress on this?

**Comment by BusyJay** [comment]
I think you still need to update `docker/release_dockerfile` too.

**Comment by Hoverbear** [comment]
:woman_facepalming: You're right.

**Pull request merged**